l1-contracts-foundry/script/Utils.sol

```diff
@@ -144,12 +154,170 @@ library Utils {
+        uint256 gasPrice = bytesToUint256(vm.rpc("eth_gasPrice", "[]"));
+
+        uint256 requiredValueToDeploy = bridgehub.l2TransactionBaseCost(
+            chainId,
+            gasPrice,
+            l2GasLimit,
+            REQUIRED_L2_GAS_PRICE_PER_PUBDATA
+        ) * 2;
+
+        L2TransactionRequestDirect memory l2TransactionRequestDirect = L2TransactionRequestDirect({
+            chainId: chainId,
+            mintValue: requiredValueToDeploy,
+            l2Contract: dstAddress,
+            l2Value: 0,
+            l2Calldata: l2Calldata,
+            l2GasLimit: l2GasLimit,
+            l2GasPerPubdataByteLimit: REQUIRED_L2_GAS_PRICE_PER_PUBDATA,
+            factoryDeps: factoryDeps,
+            refundRecipient: msg.sender
+        });
+
+        address baseTokenAddress = bridgehub.baseToken(chainId);
+        if (ADDRESS_ONE != baseTokenAddress) {
+            IERC20 baseToken = IERC20(baseTokenAddress);
+            vm.broadcast();
+            baseToken.approve(l1SharedBridgeProxy, requiredValueToDeploy);
+            requiredValueToDeploy = 0;
+        }
+
+        executeUpgrade({
+            _governor: bridgehub.owner(),
+            _salt: bytes32(0),
+            _target: l1SharedBridgeProxy,
+            _data: abi.encodeCall(bridgehub.requestL2TransactionDirect, (l2TransactionRequestDirect)),
+            _value: requiredValueToDeploy,
+            _delay: 0
+        });
+    }
+
+    /**
+     * @dev Publish bytecodes to l2 through l1
+     */
+    function publishBytecodes(
+        bytes[] memory factoryDeps,
+        uint256 chainId,
+        address bridgehubAddress,
+        address l1SharedBridgeProxy
+    ) public {
+        runL1L2Transaction({
+            l2Calldata: "",
+            l2GasLimit: MAX_PRIORITY_TX_GAS,
+            factoryDeps: factoryDeps,
+            dstAddress: 0x0000000000000000000000000000000000000000,
+            chainId: chainId,
+            bridgehubAddress: bridgehubAddress,
+            l1SharedBridgeProxy: l1SharedBridgeProxy
+        });
+    }
+
+    /**
+     * @dev Read hardhat bytecodes
+     */
+    function readHardhatBytecode(string memory artifactPath) public view returns (bytes memory) {
+        string memory root = vm.projectRoot();
+        string memory path = string.concat(root, artifactPath);
+        string memory json = vm.readFile(path);
+        bytes memory bytecode = vm.parseJsonBytes(json, ".bytecode");
+        return bytecode;
+    }
+
+    function executeUpgrade(
+        address _governor,
+        bytes32 _salt,
+        address _target,
+        bytes memory _data,
+        uint256 _value,
+        uint256 _delay
+    ) public {
+        IGovernance governance = IGovernance(_governor);
+
+        IGovernance.Call[] memory calls = new IGovernance.Call[](1);
+        calls[0] = IGovernance.Call({target: _target, value: _value, data: _data});
+
+        IGovernance.Operation memory operation = IGovernance.Operation({
```

Comment:
There is a new small feature here so we can print out the calldata if needed: 
https://github.com/matter-labs/era-contracts/blob/bffbf0a72c1ac32d6c9ebf95999efe2953f19138/l1-contracts/src.ts/deploy.ts#L473

Reply:
Let's make a bit more complex solution, not just printing call data, we could figure out what we need to do..
using foundry we can just save all txs and do not broadcast them